File: be/src/main/java/nl/zzpmatcher/business/UserManagement.java
package nl.zzpmatcher.business;

import static nl.zzpmatcher.business.utils.Validator.validate;

public class UserManagement {

    private final UserRepository userRepository;

    public UserManagement(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public User createUser(UserRegisterCommand userRegisterCommand) {
        validate(isPasswordNonEmpty(userRegisterCommand), PasswordMandatoryException::new);
        validate(arePasswordsEqual(userRegisterCommand), PasswordShouldBeEqual::new);
        validate(isEmailAddressUnique(userRegisterCommand.getEmailaddress()), DuplicateEmailAddressException::new);

        return userRepository.save(userRegisterCommand.createUser());
    }

    private boolean isPasswordNonEmpty(UserRegisterCommand userRegisterCommand) {
        return userRegisterCommand.getPassword() != null && userRegisterCommand.getPassword().length() > 0;
    }

    private boolean arePasswordsEqual(UserRegisterCommand userRegisterCommand) {
        return userRegisterCommand.getPassword().equals(userRegisterCommand.getPassword2());
    }

    private boolean isEmailAddressUnique(String emailAddress) {
        return userRepository.findUserByEmailaddress(emailAddress) == null;
    }

    public class DuplicateEmailAddressException extends RuntimeException {
        DuplicateEmailAddressException() {
            super("DuplicateEmailAddressException");
        }
    }

    class PasswordMandatoryException extends RuntimeException {
        PasswordMandatoryException() {
            super("PasswordMandatoryException");
        }
    }

    class PasswordShouldBeEqual extends RuntimeException {
        PasswordShouldBeEqual() {
            super("PasswordShouldBeEqual");
        }
    }
}

File: be/src/main/java/nl/zzpmatcher/userlogon/business/UserManagement.kt
package nl.zzpmatcher.userlogon.business

import nl.zzpmatcher.utils.Validator.validate

class UserManagement(private val userRepository: UserRepository) {

    fun createUser(userRegisterCommand: UserRegisterCommand): User {
        validate(isPasswordNonEmpty(userRegisterCommand)) { throw PasswordMandatoryException() }
        validate(arePasswordsEqual(userRegisterCommand), { throw PasswordShouldBeEqual() })
        validate(isEmailAddressUnique(userRegisterCommand.emailaddress), { throw DuplicateEmailAddressException() })

        return userRepository.save(userRegisterCommand.createUser())
    }

    private fun isPasswordNonEmpty(userRegisterCommand: UserRegisterCommand): Boolean {
        return userRegisterCommand.password != null && userRegisterCommand.password!!.length > 0
    }

    private fun arePasswordsEqual(userRegisterCommand: UserRegisterCommand): Boolean {
        return userRegisterCommand.password == userRegisterCommand.password2
    }

    private fun isEmailAddressUnique(emailAddress: String?): Boolean {
        return userRepository.findUserByEmailaddress(emailAddress!!) == null
    }

    inner class DuplicateEmailAddressException internal constructor() : RuntimeException("DuplicateEmailAddressException")

    inner class PasswordMandatoryException internal constructor() : RuntimeException("PasswordMandatoryException")

    inner class PasswordShouldBeEqual internal constructor() : RuntimeException("PasswordShouldBeEqual")
}
